pkg/service/did/resolution/resolver.go

```diff
@@ -0,0 +1,108 @@
+package resolution
+
+import (
+	"context"
+
+	didsdk "github.com/TBD54566975/ssi-sdk/did"
+	"github.com/TBD54566975/ssi-sdk/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	didint "github.com/tbd54566975/ssi-service/internal/did"
+	utilint "github.com/tbd54566975/ssi-service/internal/util"
+)
+
+type Resolver interface {
+	Resolve(ctx context.Context, did string, opts ...didsdk.ResolutionOptions) (*didsdk.ResolutionResult, error)
+}
+
+// ServiceResolver is a resolver that can resolve DIDs using a combination of local and universal resolvers.
+type ServiceResolver struct {
+	resolutionMethods []string
+	hr                Resolver
+	lr                *localResolver
+	ur                *universalResolver
+}
+
+// NewServiceResolver creates a new ServiceResolver instance which can resolve DIDs using a combination of local and
+// universal resolvers.
+func NewServiceResolver(handlerResolver Resolver, resolutionMethods []string, universalResolverURL string) (*ServiceResolver, error) {
+	if len(resolutionMethods) == 0 {
+		return nil, errors.New("no resolution methods configured")
+	}
+
+	// instantiate sdk resolver
+	sdkResolver, err := didint.BuildResolver(resolutionMethods)
+	if err != nil {
+		return nil, errors.Wrap(err, "instantiating SDK DID resolver")
+	}
+	lr, err := newLocalResolver(sdkResolver)
+	if err != nil {
+		return nil, errors.Wrap(err, "instantiating local resolver")
+	}
+
+	// instantiate universal resolver
+	var ur *universalResolver
+	if universalResolverURL != "" {
+		ur, err = newUniversalResolver(universalResolverURL)
+		if err != nil {
+			return nil, errors.Wrap(err, "instantiating universal resolver")
+		}
+	}
+
+	return &ServiceResolver{
+		resolutionMethods: resolutionMethods,
+		hr:                handlerResolver,
+		lr:                lr,
+		ur:                ur,
+	}, nil
+}
+
+// Resolve resolves a DID using a combination of local and universal resolvers. The ordering is as follows:
+// 1. Try to resolve with the handlers we have, wrapping the resulting DID in resolution result
+// 2. Try to resolve with the local resolver
+// 3. Try to resolve with the universal resolver
+func (sr *ServiceResolver) Resolve(ctx context.Context, did string, opts ...didsdk.ResolutionOptions) (*didsdk.ResolutionResult, error) {
+	// check the did is valid
+	if _, err := utilint.GetMethodForDID(did); err != nil {
+		return nil, errors.Wrap(err, "getting method DID")
+	}
+
+	ae := util.NewAppendError()
```

Comment:
When I think of errors being returned, I'm typically thinking of what callers of the method can do with them.

Given the above, curious what's the rationale for returning multiple errors in a single error? Could this be simplified by returning a single error?

Reply:
I was thinking that we may not know the exact error to propagate, if multiple resolution methods were attempted and failed. the error should be meaningful, which in this case means showing its lineage